### PR TITLE
support fetching remote state and use across deployments

### LINKED ISSUE #680: Support create folders and dynamically pass the folder id to project creations

### PR DESCRIPTION
**Pull request by xingao267**
Fixes #680  #683

### CODE REVIEW

File: internal/tfengine/tfengine.go [hunk 1]
```diff
@@ -248,6 +255,9 @@ func ConvertToLocalBackend(path string) error {
 		}
 
 		b = backendRE.ReplaceAll(b, nil)
+		b = remoteStateResourceRE.ReplaceAll(b, nil)
+		b = remoteStateDataRE.ReplaceAll(b, []byte("000"))
```
Comment (umairidris) [inline internal/tfengine/tfengine.go:259]: add comment for 000

File: templates/tfengine/components/folders/outputs.tf [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */ -}}
+
+output "folder_ids" {
+  value = {
+    {{range .folders -}}
+    "{{resourceName . "display_name"}}" = google_folder.{{resourceName . "display_name"}}.folder_id
```
Comment (umairidris) [inline templates/tfengine/components/folders/outputs.tf:18]: no need for quotes around the keys

File: templates/tfengine/components/terraform/states/main.tf [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */ -}}
+
+{{range .states -}}
+data "terraform_remote_state" "{{resourceName . "prefix"}}" {
+  backend = "gcs"
+  config = {
+    bucket = "{{$.state_bucket}}"
```
Comment (umairidris) [inline templates/tfengine/components/terraform/states/main.tf:19]: do we want to support the case where state bucket is different?

**Review by umairidris — APPROVED** [review]

File: internal/tfengine/tfengine.go (hunk 1 above)
Comment (xingao267) [reply internal/tfengine/tfengine.go:260]: Done

File: templates/tfengine/components/folders/outputs.tf (hunk 2 above)
Comment (xingao267) [reply templates/tfengine/components/folders/outputs.tf]: Done

File: templates/tfengine/components/terraform/states/main.tf (hunk 3 above)
Comment (xingao267) [reply templates/tfengine/components/terraform/states/main.tf:19]: Maybe when there is use case for that? At least for now, when using DPT, the state bucket is set at the top level config and fully managed by DPT.

**Pull request merged**